Epgsearch minor bugfixes

prl001 opened the pull request:
Minor bug fixes to fix:

- Crash when the image's `EPGSelection` doesn't inherit from `HelpableScreen`.
- Use correct serviceref as "current service" when the currently playing service had isGroup flag set.
- Call `self.recalcEntrySize()` after `self["list"].l.setList()` instead of before it, so that the scrollbar width is properly set before `self.recalcEntrySize()` asks for the list width.
- Use the parent class's list width instead of the full list width.

RobvanderDoes commented:
This gives the following crash in XTA:
[enigma2_crash_1470646333.zip](https://github.com/oe-alliance/enigma2-plugins/files/406419/enigma2_crash_1470646333.zip)
EpgSearch has been called from EventView context menu; that; s the only way, as there's no EpgSearch in S-EPG, nor in PluginBrowser.

prl001 commented:
Pressing RED within EPG Search invokes an IMDB search for the currently highlighted event. It calls `self.redButtonPressed()` which in VIX and Beyonwiz images is implemented in `EPGSelection`. XTA's `EPGSelection` doesn't implement `self.redButtonPressed()`, hence the crash when the code tries to add the method to an `ActionMap`.

`EPGSearch` has had this function on the RED button since 2011, so I it's not connected to the recent changes to `EPGSearch`. It's not too hard to fix, but it will have to wait for tomorrow morning my time. I hope that getting it to work in XTA isn't going to be too much of a saga ;)

RobvanderDoes commented:
For reasons unknown (to me) XTA has always been using a lite version of EpgSearch (see https://github.com/oe-alliance/enigma2-plugins/tree/2.3/epgsearchpli). Maybe just because of that reason?
I'm trying to get your enhanced EpgSearch into their builds, as it does have it's advantages, and in doing so I ran into the issue I presented.

BTW: You don't like a saga? http://www.saga-gis.org

prl001 commented:
Never used that one. The only one I've read (in translation) is Beowulf, and it's actually rather short :D

prl001 commented:
> For reasons unknown (to me) XTA has always been using a lite version of EpgSearch .... Maybe just because of that reason?
I'm trying to get your enhanced EpgSearch into their builds, as it does have it's advantages, and in doing so I ran into the issue I presented.

Quite likely. That explains why you've just found a bug for standard EPGSearch in XTA when the bug has actually been there for about 5 years :)

I'll see if I can get the XTA `EPGSelection` running in the Beyonwiz image. That will mean I'll be able to test it better before I add the fixes to this pull request.

prl001 commented:
Here's a test version of `EPGSearch` that should at least initialise properly in XTA. I wasn't able to get the XTA `EPGSelection` running in the Beyonwiz image, so it's not tested against the XTA environment.

The following buttons aren't expected to work in the XTA environment:
RED for IMDb search
LEFT, RIGHT, PREV, NEXT, CH+, CH- for page up/down
REC for add record timer
long-REC for add zap timer

EPG and INFO should work sensibly in the XTA environment, but they may behave slightly differently from how they behave in the Beyonwiz or VIX environment.

Other functions should work in the XTA environment.

This test version of `EPGSearch` is based on the version at the head of this branch.

[EPGSearch.zip](https://github.com/oe-alliance/enigma2-plugins/files/408122/EPGSearch.zip)

RobvanderDoes commented:
Sorry: 
[enigma2_crash_1470712438.zip](https://github.com/oe-alliance/enigma2-plugins/files/408142/enigma2_crash_1470712438.zip)

prl001 commented:
Missed one! :(

I'll post a new test version later this afternoon. Got an appointment now.

prl001 commented:
OK, I've fixed that one, found and fixed another possible problem (not handling all the cases where the AutoTimer plugin isn't available), and re-checked that all the remaining fixed entries in the `ActionMap`s were implemented.

In addition to the above limitations in the XTA image:
long-BLUE to show the AutoTimer list is disabled.
BLUE to add an AutoTimer is only available if the AutoTimer plugin is available.

E&OE :)

[EPGSearch.zip](https://github.com/oe-alliance/enigma2-plugins/files/408356/EPGSearch.zip)

RobvanderDoes commented:
The saga continues: 
[enigma2_crash_1470733999.zip](https://github.com/oe-alliance/enigma2-plugins/files/408583/enigma2_crash_1470733999.zip)

prl001 commented:
I think I might leave trying to adapt EPG Search to XTA for others. There seem to be a growing number of incompatibilities between the `EPGList` and `EPGSelection` implementations in the Beyonwiz/VIX images and in XTA. The crash in the most recent test is because XTA doesn't have the font size adjustment code that the other two have. A bit more poking around showed that XTA's handling of timer icons is also quite different.

I think that the cleanest approach would be to incorporate the search filtering and orbpos changes in epgsearch into epgsearchpli. Because I can't run the XTA `EPGList` and `EPGSelection` implementations inside the Beyonwiz image, it means that I couldn't do any testing, so I think it woud be best left to someone who can do both the coding and the testing.

I'd be happy to have this pull request merged as-is.

RobvanderDoes commented:
OK, understandable. Thanks for trying anyway.

But: using the actual pull request brings in again the old issue: when searching on a stream search results (current service/bouquet) show all the EPG from the the real service, not from the stream. I tough you would filter that out, just as in the test-version?

prl001 commented:
I think I'm running out of ways to say it. There's no way that I know of to find out whether an event in the EPG cache is from a streaming service or from a broadcast service. They always return servicerefs that look like broadcast servicerefs.

You can either have the search return results for streaming services, but have them show  the orbpost text that a broadcast service would show (the orbpos for DVB-S, or "DVB-T", or "DVB-C"), or not have results returned for streaming services at all. I don't know how anything else can be implemented.

RobvanderDoes commented:
I did some more tests, all on streams, and now I'm even more puzzled.
On some streams it now works as expected (i.e. no results on current service & bouquet, and results on all). On other streams there is only no result on current service, but on current bouquet (and on all) there is.

RobvanderDoes commented:
> I think I'm running out of ways to say it. There's no way that I know of to find out whether an event in the EPG cache is from a streaming service or from a broadcast service. They always return servicerefs that look like broadcast servicerefs.
> 
> You can either have the search return results for streaming services, but have them show the orbpost text that a broadcast service would show (the orbpos for DVB-S, or "DVB-T", or "DVB-C"), or not have results returned for streaming services at all. I don't know how anything else can be implemented.
> 
That was clear already, thanks.
But I thought the previous test results (no show search results on streams) was the way to go. Anyway: I'm happy with the other solution (do show results, but with 'wrong' orb pos) as well.
Anyway: quite ambiguous results  now, as you see in my last report.

prl001 commented:
Do all the streams you're testing on have corresponding broadcast streams with up-to-date EPG data?

Do streaming services carry their own EIT EPG data?

RobvanderDoes commented:
Yes on both.
Actually EIT is the only EPG-data that finds it's way to the client box via streams. And my client boxes have no other EPG imports activated.

prl001 commented:
> But I thought the previous test results (no show search results on streams) was the way to go.

Sorry, I missed that that result was what you preferred.

> Anyway: I'm happy with the other solution (do show results, but with 'wrong' orb pos) as well.

Both solutions are easy to achieve. I just need to know which one is wanted. If the second one is wanted, then the ambiguous results above just "softly and suddenly vanish away".

RobvanderDoes commented:
OK, clear: I'll merge this then. Thanks again.

prl001 commented:
> Yes on both.
Actually EIT is the only EPG-data that finds it's way to the client box via streams. And my client boxes have no other EPG imports activated.

I'm happy to add back the debug code and have a look and try to work out why the filtering is sometimes failing to return results.

RobvanderDoes commented:
Well, when you've done all the changes you intended to do (I know: a bit more as well), I'm happy to do some more testing as well. Maybe not today though, but there seems to be no hurry, as the basis is sound now.

prl001 commented:
I've attached a version of the current EPGSearch.py with debug code back in.

If you could test with a search on stream services that fails when it should return matches and post the debug log that would be great.

I think I've found a way of doing the filtering that will correctly preserve the servicerefs of matches on streaming services. It will be a little slower than the current code, but it should be OK. I haven't started coding/testing yet.

RobvanderDoes commented:
Thanks. 
Attached the log of 4 searches on 1 service: all show results, except the search on 'current bouquet'.
[Enigma2_debug_2016-08-11_05-24-42.zip](https://github.com/oe-alliance/enigma2-plugins/files/412497/Enigma2_debug_2016-08-11_05-24-42.zip)

prl001 commented:
Thanks.

Weird. When the code builds the set of servicerefs in the bouquet (which I think is your "Opname kanalen" favourites list, userbouquet.favourites.tv), the serviceref for 1:0:1:1B03:3FE:1:C00000:0:0:0: is missing from the filter. Do I have the correct favouries list?

Here's another EPGSearch.py with some more debugging. I just need the result for the search against the current bouquet that failed in the previous test.

[EPGSearch.zip](https://github.com/oe-alliance/enigma2-plugins/files/412758/EPGSearch.zip)

RobvanderDoes commented:
Here you are:
[Enigma2_debug_2016-08-11_11-44-06.zip](https://github.com/oe-alliance/enigma2-plugins/files/412903/Enigma2_debug_2016-08-11_11-44-06.zip)

RobvanderDoes commented:
Sorry forgot to answer your question :(
The stream I searched on is 'TV5Monde Europe' in the bouquet 'Opname kanalen (remote of 192.168.1.75)(TV)', which is the remote version of 'Opname kanalen' in which are the same services.

prl001 commented:
OK, I think I've found the problem. TV5Monde Europe' in  'Opname kanalen (remote of 192.168.1.75)(TV)' is set up with an alternative, and alternatives aren't handled correctly in the filter setup.

I missed that because as far as I can see the old Settings.zip doesn't have any use of alternatives, though anyway, overnight I'd been wondering about how alternatives should be handled.

I don't think it will be hard to fix.

prl001 commented:
OK, here's a version that should handle alternatives correctly, and i should search correctly now for events on 'TV5Monde Europe' in a "current bouquet" search when the current bouquet is 'Opname kanalen (remote of 192.168.1.75)(TV)' . It still has the debug code in it in case it doesn't work properly.

[EPGSearch.zip](https://github.com/oe-alliance/enigma2-plugins/files/414511/EPGSearch.zip)

RobvanderDoes commented:
This seems to be working flawless. I tried all 4 searches on many services, real and remote, and all results were as expected.
Searches on all bouquets/all services are considerably slower than before, but not unreasonable (given the fact that I have many bouquets and many services in my settings).
So it's looking good :+1: 

PS: Normally I don't have any alternatives; I never use them. If you found any in my previous test, that must have been a left over of other tests. Anyway: I first ran tests using the very same settings as before, so if there was an alternative (TV5 monde), it now works fine as well. After that I changed my settings to the normal situation (i.e. no alternatives) and all was well as well.

prl001 commented:
> Searches on all bouquets/all services are considerably slower than before

They shouldn't be slower by any noticeable amount. The added code to check for and use alternatives shouldn't slow things down by a significant amount. The code that does "all services" searches is essentially unchanged (even from before I started working on the "Search scope" additions).

I'll check in the fixes that deal with alternatives properly, and go back to finish the code to handle orbpos properly on streaming services.

I'll also have a look at making the "\x86"/"\x87" translations work properly.

RobvanderDoes commented:
> > Searches on all bouquets/all services are considerably slower than before
> 
> 
> They shouldn't be slower by any noticeable amount. 
> 

Actually it's the search on 'all bouquets' that is (or seems to be) a bit slower, not the 'all services' search. But that's no problem.
